Do not allow editing username via User API

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-12586

There is a bug in the User API that allows API users to update a User object's username which we definitely do not support.

In theory this impacts any non-editable key, not just the `username` field, due to the catch-all `else` at the end of the `_update` block. This includes the `_id` field. To address this I created the `valid_keys` list to ensure we only update keys we intend to be mutable. 

~~Additionally, the `_update` call is used in both the `obj_update` and `obj_create` methods. Since the `obj_create` method's bundle contains the `username`, I added a parameter `raise_error_if_invalid_key` (verbose...I know) which as it says, will raise an error if there is an invalid key supplied vs silently skipping that key.~~
Update: I've instead popped off values that shouldn't be passed to the update method after object creation.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Tested locally, and ensured user creation is still successful as well.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added test for the username bug in particular.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
disallow updating username; add test to confirm behavior

snopoke commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -312,7 +319,7 @@ def obj_create(self, bundle, **kwargs):
             )
             del bundle.data['password']
-            self._update(bundle)
+            self._update(bundle, raise_error_if_invalid_key=False)
```
Is `bundle.data` just a dict? We could copy it and then pop off the domain, username, password and email keys before passing it to `_update`.

That would then raise an exception if there were additional keys in the data which I think is better than silently ignoring them.

gherceg replied:
I didn't copy the dict since it isn't used after this call and that is how the `password` field is handled currently. I also didn't include the email and domain fields because the _update method has logic to account for when a value hasn't changed, and I don't think there is a domain field in this dictionary.

gherceg committed:
pop username and password _update in obj_create

mkangia commented on corehq/apps/api/tests/user_resources.py:
```diff
@@ -168,6 +168,21 @@ def test_create(self):
+                                                   method='PUT')
+        unmodified_user = CommCareUser.get(user._id)
+        self.assertEqual(unmodified_user.username, 'test')
+        self.assertEqual(response.status_code, 400)
```
nit: good to test the message in response as well.

mkangia commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -264,7 +269,7 @@ def _update(self, bundle, user_change_logger=None):
                         user_change_logger.add_info(UserChangeMessage.groups_info(groups))
                     should_save = True
-                elif key in ['email', 'username']:
+                elif key == 'email':
```
Nice!
good to see that work for user change logger helped point these things out, https://dimagi-dev.atlassian.net/browse/SUPPORT-10484.

mkangia commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -311,7 +316,10 @@ def obj_create(self, bundle, **kwargs):
             )
-            del bundle.data['password']
+            # username cannot be updated
+            bundle.data.pop('username', None)
```
will this ever get hit? if there is username in bundle.data, there will be an error already.

gherceg replied:
This is in the `obj_create` method so there should be a `username` key. Definitely not my preferred way to handle this but was just making minimal changes for the time being.

mkangia commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -222,7 +222,12 @@ def get_resource_uri(self, bundle_or_obj=None, url_name='api_dispatch_detail'):
     def _update(self, bundle, user_change_logger=None):
         should_save = False
+        valid_keys = [
+            'phone_numbers', 'groups', 'email', 'password', 'user_data', 'first_name', 'last_name', 'language'
```
should this on the model? I am thinking if the same set of fields should be allowed for update in api and HQ UI as well, so this should be central and not API specific.

gherceg replied:
Yeah I agree it would be wise to centralize that logic. I'll look into that.

gherceg replied:
fb5bb18dc4b515f61911bfd3e5806c3f1a8693a2 is not what Manish had in mind, but I think unifying the fields that are editable via the UI vs the api would require work that is outside the scope of this ticket. I felt moving the list of valid keys to the user model at least made it more visible. @snopoke do you have any opinions on this?

snopoke replied:
+1 on unifying the editable fields being out of the scope of this PR.

I think if you put them like this on the user class thy will also get added to the JSON output and saved in the DB. I think just making it a constant would be sufficient.

esoergel replied:
I actually think it makes sense to store on the resource class itself, as the set of editable fields could vary from version to version.  Like, if we make a v0.6 for this API, we might want to disallow editing your password.

Also, doesn't tastypie do this for you?  I see most of the fields in this API are defined here:
https://github.com/dimagi/commcare-hq/blob/ac088c211bfee503eb3011b0aab649fb893009c9/corehq/apps/api/resources/v0_1.py#L25-L33
You should be able to pass `readonly=True` to a field to prevent users from writing to it.  It looks like the `WebUserResource` and the older versions of both only allow `GET` requests, so it shouldn't affect those.

gherceg replied:
Offlined with Ethan but never updated on this thread. The readonly attribute does handle this for you if using `hydrate` _and_ the django ORM. I put up a solution b3d439f4e523597a112160c9daffd0dcd44e2e4d that still makes use of the readonly attribute, but in a more custom way. More on that below.

gherceg committed:
assert response message is correct

mkangia commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -311,7 +316,10 @@ def obj_create(self, bundle, **kwargs):
+            # username cannot be updated
+            bundle.data.pop('username', None)
+            # password was just set
+            bundle.data.pop('password', None)
```
just curious about this change. Using pop is better than del?

snopoke replied:
I prefer `pop` since you can pass a default value to avoid raising an exception if the key does not exist. With `del` you have to guard against a `KeyError` exception.

gherceg committed:
add api_editable_keys list to CommCareUser class

gherceg committed:
move to const.py file in api app

gherceg committed:
Merge branch 'master' into gh/bugs/user-api-username

gherceg committed:
replace const list with readonly check on user resource

gherceg commented:
I took @esoergel's suggestion to use the readonly attribute as this seemed like a nice way to handle this, though I don't think we will be able to use it in the traditional sense. I think typically this would be checked in `hydrate`, but even after implementing that method I wasn't seeing it get called prior to `obj_update`. Inspired by [this code](https://github.com/toastdriven/django-tastypie/blob/v0.9.14/tastypie/resources.py#L879-L881), we can still make use of the `readonly` attribute. 

The implementation now ensures that only attributes that are not readonly _and_ specified in the CommCareUserResource are editable. This covers edge cases like `_id` and other outliers since our `_update` method has a [catch all](https://github.com/dimagi/commcare-hq/blob/b3d439f4e523597a112160c9daffd0dcd44e2e4d/corehq/apps/api/resources/v0_5.py#L297-L302). It seems like changing that catch all might be a good idea as well.

gherceg committed:
add language to user resource; add exception for password

gherceg committed:
fix failing test

gherceg committed:
fix other failing test

gherceg commented:
I'll be curious to hear what others think on Monday, but at the moment I'm leaning towards reverting back to storing a list of keys on the resource that we need to manually update when needed, and is disjointed from the actual couch user (which the current solution still is as well).

snopoke commented:
> I'll be curious to hear what others think on Monday, but at the moment I'm leaning towards reverting back to storing a list of keys on the resource that we need to manually update when needed, and is disjointed from the actual couch user (which the current solution still is as well).

I do like the simplicity and clarity of this approach so I'm fine with going back to it.

esoergel reviewed (approved):
Gotcha, that's a shame it doesn't work as smoothly as I hoped, but I guess that's been pretty standard with tastypie - the easy stuff is very easy, and anything unusual gets real messy fast.  Also sounds fine to go to an explicit blacklist on the resource as you suggested.

gherceg committed:
revert back to simple immutable fields

gherceg committed:
remove catch-all in favor of explicit keys to update

gherceg commented:
I added an `immutable_fields` property (ea6c6493b56f2fc087eff59d53038aedb83c09e8) to the CommCareUserResource. Notably we didn't add the ability to edit mobile users until v0.5, so it is safe to add it there. If the json input contains an immutable key, a BadRequest error is saved and any pending updates are not saved to the user.

Additionally, I removed the catch-all `else` block (f903a1fea3e1c948fd2e54e0d4ce5818a934cc9e) in favor of checking a list of keys that are convenient to group together since they can be updated in the same way. I'd just look at the files changed at this point since this has a messy commit history.

esoergel reviewed (approved):
Thanks for sticking with this.  It's wild how much free rein some of this old code used to give.

snopoke commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -293,12 +296,13 @@ def _update(self, bundle, user_change_logger=None):
                     setattr(bundle.obj, key, value)
                     should_save = True
+                else:
+                    raise BadRequest(f'Attempted to update unknown key {key}.')
```
nit: better wording

```suggestion
                    raise BadRequest(f'Attempted to update unknown field {key}.')
```

snopoke reviewed (approved):
LGTM

gherceg committed:
Update exception message wording

Co-authored-by: Simon Kelly <skelly@dimagi.com>

gherceg committed:
fix update unknown key test
